### PR TITLE
Add scalar division methods

### PR DESCRIPTION
**Pull request by d-torrance**
I was using M2 to double-check some computations while grading calculus homework and was surprised to get the following error:

```m2
i1 : vector {1, 2, 3} / 4
stdio:1:17:(3): error: no method for binary operator / applied to objects:
                              3
            | 1 | (of class ZZ )
            | 2 |
            | 3 |
      /     4 (of class ZZ)
```

We add scalar division methods for vectors and matrices by numbers and ring elements.  So afterward:

```m2
i1 : vector {1, 2, 3} / 4

o1 = | 1/4 |
     | 1/2 |
     | 3/4 |

       3
o1 : QQ
```

### CODE REVIEW

File: M2/Macaulay2/m2/matrix.m2 [hunk 1]
```diff
@@ -85,6 +85,8 @@ Matrix * Number :=
 Matrix * RingElement := (m,r) -> (
     if ring r =!= ring m then try r = promote(r,ring m) else m = promote(m,ring r);
      map(target m, source m, reduce(target m, raw m * raw r)))
+Matrix / Number      :=
+Matrix / RingElement := (m,r) -> 1/r * m
```
Comment (mahrud) [inline M2/Macaulay2/m2/matrix.m2:89]: I think this is intentionally not defined:
```m2
i3 : R = kk[x,y]

o3 = R

o3 : PolynomialRing

i4 : matrix{{x}} / x

o4 = | 1 |

                    1             1
o4 : Matrix (frac R)  <-- (frac R)
```

Comment (mahrud) [reply M2/Macaulay2/m2/matrix.m2:89]: But this is defined:
```m2
i3 : matrix{{x}} // x

o3 = | 1 |

             1      1
o3 : Matrix R  <-- R
```

Comment (mahrud) [reply M2/Macaulay2/m2/matrix.m2:89]: Even over integers:
```m2
i2 : Matrix / Number      := (m,r) -> 1/r * m

o2 = FunctionClosure[stdio:2:30-2:39]

o2 : FunctionClosure

i3 : matrix{{2}} / 2

o3 = | 1 |

              1       1
o3 : Matrix QQ  <-- QQ
```

Comment (d-torrance) [reply M2/Macaulay2/m2/matrix.m2:89]: I'm proposing that we define them.  :)  I think the result being over the fraction field seems like reasonable behavior.  That's how `ZZ / ZZ` and `RingElement / RingElement` work.

Comment (mahrud) [reply M2/Macaulay2/m2/matrix.m2:89]: I guess my point is that using `1/r * m` instead makes it clear that you _should_ expect something over the fraction field, so maybe at least caveat this in the documentation.

**Comment by pzinn** [comment]
I think it was intentionally undefined. In linear algebra, we learn that we can takes "scalar * vector", but not the other way around.

**Comment by pzinn** [comment]
actually, I'm confused, multiplication already works -- `matrix{{3}}*3` or `vector{3}*3`...

**Comment by pzinn** [comment]
never mind my previous comment. then I guess there's no reason not to have division work as well.

**Comment by mahrud** [comment]
Actually, how this behaves in the noncommutative case is a great point. There are two _different_ methods for scalar multiplication on left and right on purpose:
https://github.com/Macaulay2/M2/blob/1fd238210e9a37ddaec3b700ccd22f06ff5f28f5/M2/Macaulay2/d/engine.dd#L177-L188
The `opposite_mult` argument (whether or not it's actually honored in the engine) distinguishes them:
https://github.com/Macaulay2/M2/blob/1fd238210e9a37ddaec3b700ccd22f06ff5f28f5/M2/Macaulay2/e/matrix.hpp#L117
So at the very least division should be defined as `m * (1/r)` not the other way around.

**Comment by pzinn** [comment]
ah, yes, cf #2811

**Comment by d-torrance** [comment]
Makes sense!  I've updated it to use right multiplication.

File: M2/Macaulay2/packages/Macaulay2Doc/doc.m2 [hunk 2]
```diff
@@ -403,6 +403,33 @@ document {
...
+  Description
+    Text
+      This operation is equivalent to right scalar multiplication by the
+      multiplicative inverse of @CODE "c"@, i.e., @CODE "v * (1/c)"@.
+    Example
+      R = QQ[a,b,c,d]
+      matrix {{d, -b}, {-c, a}} / (a * d - b * c)
+  Caveat
+    The base ring of the output will be the fraction field of the base ring
+    of @CODE "v"@.
```
Comment (pzinn) [inline M2/Macaulay2/packages/Macaulay2Doc/doc.m2:430]: I'm not sure I agree with this "Caveat". For example, `matrix{{1}}/2.` will return a matrix in `RR`...

Comment (d-torrance) [reply M2/Macaulay2/packages/Macaulay2Doc/doc.m2:430]: Hmm, good point.  Maybe something more general like "the base ring of the output will be a field containing the base ring of v"?

Comment (d-torrance) [reply M2/Macaulay2/packages/Macaulay2Doc/doc.m2:430]: I went ahead and made this change

**Comment by d-torrance** [comment]
@pzinn , @mahrud: Any additional thoughts on this PR?  Is this okay to merge?

**Review by mahrud — APPROVED** [review]

**Pull request merged**